### DISCUSSION TITLE
Error: Can't connect to Chrome DevTools! The @wdio/devtools-service currently only supports Chrome, Chromium and Firefox!

### DISCUSSION THREAD

**Discussion by Sandip-lab007** [webdriverio/webdriverio#11883] [Q&A]
Hi All,
Error: Can't connect to Chrome DevTools! The @wdio/devtools-service currently only supports Chrome, Chromium and Firefox!

Getting this error when i am trying to use CDP using wdio devtools service on selenium/standalone chrome server. but this exact setup works fine when i run it locally . can someone please help here as this exact setup was working till 5 th Dec but from 7th Dec onwards it started failing with above.
Please direct me to any resource you have.

**Comment by christian-bromann** [comment]
Please provide error logs or a reproducible example.

**Comment by th0car** [comment]
Experiencing the same problem using Selenium Grid 4. Problems started beginning of december.
Selenium Grid is hosted in an Openshift namespace and installed via Helm Chart from SeleniumHQ.
I will get back next week with logs from both Selenium Chrome pods and WDIO.

**Comment by th0car** [comment]
I'm using a WDIO project to perform a geosort test using Devtools. A Selenium grid installed using a Helm chart on an openshift project is also used.
Tests are run on two different Selenium Grid versions:
4.15.0-20231129 (chrome 119.0.6045.199 linux) 
4.16.1-20231212 (chrome 120.0.6099.71 linux)

When using Selenium Grid 4.15.0 tests run successfully as they have with different older WebdriverIO and Selenium Grid versions since the tests were implemented almost two years ago.

When using Selenium Grid 4.16.1 tests fail with the following error message:
```
[chrome 120.0.6099.71 linux #0-0] Can't connect to Chrome DevTools! The @wdio/devtools-service currently only supports Chrome, Chromium and Firefox!
Given that cloud vendors don't expose access to the Chrome DevTools Protocol this service also usually only works when running tests locally or through a Selenium Grid (https://www.selenium.dev/documentation/grid/) v4 or higher.
```

Not sure taht this is a Selenium problem or a WebdriverIO problem, but I guess the error message is from wdio. Let me know if you need me to add this as an issue instead or if it should be directed to SeleniumHQ.

WebdriverIO config:
```
export const config = {
    runner: 'local',
    hostname: 'grid.openshift.domain.com',
    port: 443,
    path: '/wd/hub',
    protocol: "https",
    specs: ['./test/specs/**/*.js'],
    maxInstances: 1,
    capabilities: [{
        browserName: 'chrome',
        platformName: "LINUX",
        "goog:chromeOptions": {
          args: [
             "--ignore-certificate-errors",
           ],
        },
    }],
    logLevel: 'info',
    bail: 0,
    baseUrl: 'https://www.domain.com/',
    waitforTimeout: 10000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 3,
    services: ['devtools'],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    },
}
```


```
{
  "name": "wdio_devtools_grid",
  "type": "module",
  "devDependencies": {
    "@wdio/cli": "^8.27.1",
    "@wdio/devtools-service": "^8.27.0",
    "@wdio/local-runner": "^8.27.0",
    "@wdio/mocha-framework": "^8.27.0",
    "@wdio/spec-reporter": "^8.27.0"
  },
  "scripts": {
    "wdio": "wdio run ./wdio.conf.js"
  }
}
```


```
import { expect, browser, $ } from "@wdio/globals";

describe("Geo sort applied", () => {
  it("can be sorted using emulated ", async () => {
    await browser.pause(5000)
    await browser.cdp("Emulation", "setGeolocationOverride", {
      latitude: 59.372133,
      longitude: 17.965764,
      accuracy: 1,
    });
    await browser.url("/page-with-geosorting");
    await browser.pause(5000)
    expect(await browser.getTitle()).toBe("Page With Geo Sorting")
  });
});
```

**Comment by christian-bromann** [comment]
> Not sure taht this is a Selenium problem or a WebdriverIO problem, but I guess the error message is from wdio. Let me know if you need me to add this as an issue instead or if it should be directed to SeleniumHQ.

This does not seem like a bug but rather the notification that for given environment the devtools protocol is not supported. Note that the browser exposes the protocol on a different port and if port forwarding is not configured correctly WebdriverIO will fail to connect.

**Comment by cir94** [comment]
Hey all,

I'm also experiencing the same issue when using wdio/devtools-service, specifically when trying to test on Firefox. Chrome, however, works fine. The error I get is this:

```
`Error: Can't connect to Chrome DevTools! The @wdio/devtools-service currently only supports Chrome, Chromium and Firefox!

Given that cloud vendors don't expose access to the Chrome DevTools Protocol this service also usually only works when running tests locally or through a Selenium Grid (https://www.selenium.dev/documentation/grid/) v4 or higher.`
``` 

The code I use which causes this error is this:

```import { envs } from '../url-module.js';

describe('UI Testing Playground - Overlapped Element', function() {
    it('should scroll to fully display name  text box, then write text', async function() {
        envs.uiTestingUrl.pathname = '/overlapped';
        await browser.url(`${envs.uiTestingUrl}`);
        await browser.emulateDevice('iPhone 12');
        const nameInput = browser.$('#name');
        await nameInput.scrollIntoView();
        await nameInput.setValue('Test');
    });
});
```

All other tests run fine except for this one on Firefox, unless it uses anything devtools related.

My config looks like this:

```export const config = {
    specs: ['../tests/**'],
    maxInstances: 10,
    maxInstancesPerCapability: 10,
    injectGlobals: true,
    capabilities: [
        {
            browserName: 'chrome',
            'goog:chromeOptions': {
                args: ['-headless', 'disable-gpu'],
            },
        },
        {
            browserName: 'firefox',
            'moz:firefoxOptions': {
                args: ['-headless'],
            },
        },
    ],
    logLevel: 'info',
    baseUrl: 'http://localhost:8080',
    bail: 0,
    waitforTimeout: 10000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 3,
    services: ['devtools'],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    }}
```

Pointers to any sort of answer or resource to read would be greatly appreciated. I've been scratching my head over this the past couple of days.

**Reply by christian-bromann** [reply]
I ultimately would not advice to use Firefox with the `@wdio/devtools-protocol`. Even though it is technically possible the Firefox team does not actively maintain this anymore. We might have some similar capabilities cross browser in the future with WebDriver. For now, I recommend not to run these test in Firefox. In fact it makes no sense to emulate an iPhone 12 on Firefox.

**Reply by cir94** [reply]
Thanks for your message, and I had no idea they weren't actively maintaining this anymore, thanks for the heads-up. I've been testing on chrome as well, and this was more of a passing curiosity as to why it wasn't working.

As for testing for an iPhone 12 on Firefox, that was just to pick a responsive mobile option it has available when using mobile mode, not so much for the sake of actually testing it.

Thanks again for your help! Appreciate it.
